src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11041,12 +11101,22 @@ private LambdaExpressionSyntax ParseLambdaExpression(SyntaxToken asyncToken)
             else
             {
                 var name = this.ParseIdentifierToken();
-                var arrow = this.EatToken(SyntaxKind.EqualsGreaterThanToken);
-                arrow = CheckFeatureAvailability(arrow, MessageID.IDS_FeatureLambda);
-
+                var exclamation = this.CurrentToken.Kind == SyntaxKind.ExclamationToken ? this.EatToken(SyntaxKind.ExclamationToken) : null;
+                SyntaxToken arrow;
+                if (this.CurrentToken.Kind == SyntaxKind.ExclamationEqualsToken && this.PeekToken(1).Kind == SyntaxKind.GreaterThanToken)
+                {
+                    var notEq = this.EatToken(SyntaxKind.ExclamationEqualsToken);
+                    arrow = ConvertToMissingWithTrailingTrivia(notEq, SyntaxKind.EqualsGreaterThanToken);
+                    arrow = AddError(arrow, ErrorCode.ERR_NeedSpaceBetweenExclamationAndEquals);
+                }
```

Comment:
1. shouldn't you make the exclamation into a missing token, like you did in the previous error tolerance case far above?
2. perhaps possible to share some code here with the above (though the benefits might be so minor).  but that would have helped keep things consistent :)